File: house-robber-ii/house-robber-ii.cpp
class Solution
{
public:
    int rob(vector<int> &nums)
    {
        int n = nums.size();
        if (n == 0)
            return 0;
        if (n == 1)
            return nums[0];
        if (n == 2)
            return max(nums[0], nums[1]);
        return max(rob(nums, 0, n - 2), rob(nums, 1, n - 1));
    }
    int rob(vector<int> &nums, int l, int r)
    {
        int prev_prev = 0, prev = 0;
        for (int i = l; i <= r; i++)
        {
            int temp = prev;
            prev = max(prev_prev + nums[i], prev);
            prev_prev = temp;
        }
        return prev;
    }
};

File: others/house-robber-ii/house-robber-ii.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if not nums:
            return 0
        if (len(nums) == 1):
            return nums[0]
        
        def go(num):
            last, now = 0, 0
            for n in num:
                last, now = now, max(last + n, now)
            return max(last, now)
        return max(go(nums[1:]), go(nums[:-1]))
    